src/finchlite/autoschedule/einsum.py

```diff
@@ -0,0 +1,219 @@
+
+    def rename_einsum(
+        self,
+        einsum: ein.Einsum,
+        new_alias: ein.Alias,
+        definitions: dict[str, ein.Einsum],
+    ) -> ein.Einsum:
+        definitions[new_alias.name] = einsum
+        return ein.Einsum(einsum.op, new_alias, einsum.idxs, einsum.arg)
+
+    def reorder_einsum(
+        self, einsum: ein.Einsum, idxs: tuple[ein.Index, ...]
+    ) -> ein.Einsum:
+        return ein.Einsum(einsum.op, einsum.tns, idxs, einsum.arg)
+
+    def compile_plan(
+        self, plan: Plan, bindings: dict[str, Any], definitions: dict[str, ein.Einsum]
+    ) -> ein.Plan:
+        bodies: list[ein.EinsumNode] = []
+
+        for body in plan.bodies:
+            match body:
+                case Plan(_):
+                    inner_plan = self.compile_plan(body, bindings, definitions)
+                    bodies.extend(inner_plan.bodies)
+                    break
+                case Query(Alias(name), Table(Literal(val), _)) if isinstance(
+                    val, Scalar
+                ):
+                    bindings[name] = val.val
+                case Query(Alias(name), Table(Literal(tns), _)) if isinstance(
+                    tns, Tensor
+                ):
+                    bindings[name] = (
+                        tns.to_numpy() if hasattr(tns, "to_numpy") else np.asarray(tns)
+                    )  # type: ignore[attr-defined]
+                case Query(Alias(name), rhs):
+                    bodies.append(
+                        self.rename_einsum(
+                            self.lower_to_einsum(rhs, bodies, bindings, definitions),
+                            ein.Alias(name),
+                            definitions,
+                        )
+                    )
+                case Produces(args):
+                    returnValues = []
+                    for arg in args:
+                        if isinstance(arg, Alias):
+                            returnValues.append(ein.Alias(arg.name))
+                        else:
+                            einsum = self.rename_einsum(
+                                self.lower_to_einsum(
+                                    arg, bodies, bindings, definitions
+                                ),
+                                self.get_next_alias(),
+                                definitions,
+                            )
+                            bodies.append(einsum)
+                            returnValues.append(einsum.tns)
+
+                    bodies.append(ein.Produces(tuple(returnValues)))
+                case _:
+                    bodies.append(
+                        self.rename_einsum(
+                            self.lower_to_einsum(body, bodies, bindings, definitions),
+                            self.get_next_alias(),
+                            definitions,
+                        )
+                    )
+
+        return ein.Plan(tuple(bodies))
+
+    def lower_to_einsum(
```

Comment:
Inline these by wrapping the pattern in each case inside of the rhs of a query node

Reply:
I think this is the old code.